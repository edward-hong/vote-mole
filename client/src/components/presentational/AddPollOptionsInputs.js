import React from 'react'
import { Field } from 'redux-form'
import styled from 'styled-components'
import { and, not, inc } from 'ramda'

import AddPollExtraOptionInput from './AddPollExtraOptionInput'
import ErrorMessage from './styled/ErrorMessage'
import BlackButton from './styled/BlackButton'
import mapIndexed from '../../utils/mapIndexed'

const PollOptionsWrapper = styled.div`
	text-align: center;
`

const AddOptionButton = styled(BlackButton)`
	margin-bottom: 10px;
`

const AddPollOptionsInputs = ({ fields, meta: { error, pristine } }) => {
	const handleAddOption = () => fields.push()

	return (
		<PollOptionsWrapper>
			<AddOptionButton outline onClick={handleAddOption}>
				Add Option
			</AddOptionButton>
			{and(and(not(pristine), error), <ErrorMessage message={error} />)}
			{mapIndexed(
				(field, index) => (
					<Field
						key={index}
						name={field}
						component={AddPollExtraOptionInput}
						deleteOption={fields.remove}
						index={index}
						placeholder={`Option ${inc(index)}`}
					/>
				),
				fields
			)}
		</PollOptionsWrapper>
	)
}

export default AddPollOptionsInputs
