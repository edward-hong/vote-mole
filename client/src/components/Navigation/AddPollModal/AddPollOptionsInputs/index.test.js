import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { reduxForm } from 'redux-form'

import AddPollOptionsInputs from '.'
import renderWithRedux from '../../../../utils/test/renderWithRedux'

describe('AddPollOptionsInputs', () => {
	it('shows error message when there is an error', () => {
		const error = 'Please fill this field'
		const { getByText } = render(
			<AddPollOptionsInputs meta={{ pristine: false, error }} fields={[]} />,
		)

		expect(getByText(error)).toBeInTheDocument()
	})

	it('renders correctly', () => {
		const EnhancedComponent = reduxForm({ form: 'add_poll' })(
			AddPollOptionsInputs,
		)
		const { container } = renderWithRedux(
			<EnhancedComponent
				meta={{ pristine: true, error: false }}
				fields={['test1', 'test2']}
			/>,
		)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('calls the add option handler when the add option button is clicked', () => {
		const EnhancedComponent = reduxForm({ form: 'add_poll' })(
			AddPollOptionsInputs,
		)
		const fields = ['test1', 'test2']
		fields.push = jest.fn()
		const { getByText } = renderWithRedux(
			<EnhancedComponent
				meta={{ pristine: true, error: false }}
				fields={fields}
			/>,
		)
		const addOptionButton = getByText('Add Option')

		fireEvent.click(addOptionButton)

		expect(fields.push).toHaveBeenCalled()
	})
})
