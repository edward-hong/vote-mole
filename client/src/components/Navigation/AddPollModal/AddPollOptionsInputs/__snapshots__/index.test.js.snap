// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPollOptionsInputs renders correctly 1`] = `
.c2 {
  margin-bottom: 3px;
}

.c2 input {
  height: 47px;
}

.c0 {
  text-align: center;
}

.c1 {
  color: #21252a;
  border-color: #21252a;
  margin-bottom: 10px;
}

.c1:hover,
.c1:active {
  background-color: #21252a;
  border-color: #21252a;
}

.c1:focus {
  box-shadow: 0 0 0 3px rgba(33,37,41,0.5);
}

<div
  class="c0"
>
  <button
    class="c1 btn btn-outline-secondary"
    type="button"
  >
    Add Option
  </button>
  <div
    class="c2 input-group"
  >
    <input
      class="form-control"
      id="test1"
      name="test1"
      placeholder="Option 1"
      type="text"
      value=""
    />
    <div
      class="input-group-append"
    >
      <button
        class="btn btn-outline-danger"
        type="button"
      >
        <span
          aria-hidden="true"
          class="fa fa-close fa-2x"
        />
      </button>
    </div>
  </div>
  <div
    class="c2 input-group"
  >
    <input
      class="form-control"
      id="test2"
      name="test2"
      placeholder="Option 2"
      type="text"
      value=""
    />
    <div
      class="input-group-append"
    >
      <button
        class="btn btn-outline-danger"
        type="button"
      >
        <span
          aria-hidden="true"
          class="fa fa-close fa-2x"
        />
      </button>
    </div>
  </div>
</div>
`;
